Start soulbinder

ferriarnus opened the pull request:
Soul Binder and recipes

# Description

Please write a summary of the changes you have implemented; explaining your design choices if necessary. Please also include a list of things to test if it isn't clear.
Please also disclose whether this is a breaking change.

# Todo <!-- Remove this section if you're submitting an already-complete PR -->

- [x] powered spawner + recipe
- [x] model
- [x] gui

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

ferriarnus committed:
Start soulbinder

TODO
-powered spawner + recipe
-model
-gui

ferriarnus committed:
Working spawner

Spawner works, but needs gui and config changes.
Soul Binder needs a gui as well.

ferriarnus committed:
remove spawnrules and add range

ferriarnus committed:
add experience

ferriarnus committed:
added XP bar

Now this also as an xp Bar
Still needs some love in the gui, like the range indicator for spawners

ferriarnus committed:
Fix loot table

-fix loot table for cap drop
-remove unneeded info from table
- made serialize itemstack by hand

Bugs/future:
-consume energy
-fix gui and tooltip
-check and fix other machine loot
-check and fix other itemstack entity storage (maybe)

ferriarnus committed:
add experience, gui and on click

ferriarnus committed:
Merge branch '1.19.x' into feature/soulbinder

ferriarnus committed:
Update to latest

also add the range indicator to the spawner.

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/SoulBinderBlockEntity.java:
```diff
@@ -0,0 +1,138 @@
+        return MachineInventoryLayout.builder()
+            .setStackLimit(1)
+            .inputSlot((slot, stack) -> stack.getCapability(EIOCapabilities.ENTITY_STORAGE).isPresent())
+            .inputSlot()
```
use slotaccess

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/task/SpawnTask.java:
```diff
@@ -0,0 +1,133 @@
+    public boolean isAreaClear() {
+        AABB range = new AABB(blockEntity.getBlockPos()).inflate(blockEntity.getRange());
+        List<? extends Entity> entities = blockEntity.getLevel().getEntities(blockEntity.getEntityType(), range, p -> true);
+        if (entities.size() >= maxEntities) { //TODO config? Max amount of entities.
```
yep, move to config

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/task/SpawnTask.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    public boolean isAreaClear() {
+        AABB range = new AABB(blockEntity.getBlockPos()).inflate(blockEntity.getRange());
+        List<? extends Entity> entities = blockEntity.getLevel().getEntities(blockEntity.getEntityType(), range, p -> true);
```
filter to livingentity, as the spawning should not be stopped by fishingbobber, partentities or items

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/task/SpawnTask.java:
```diff
@@ -0,0 +1,133 @@
+        if (entities.size() >= maxEntities) { //TODO config? Max amount of entities.
+            return false;
+        }
+        if (BlockPos.betweenClosedStream(range).filter(pos -> blockEntity.getLevel().getBlockEntity(pos) instanceof PoweredSpawnerBlockEntity).count() >= maxSpawners) { //TODO config? Max amount of spawners.
```
why should there be a limit for this?

ferriarnus replied:
There was one in the past, I guess to stop massive farms?

justliliandev replied:
If we add a limit it should be discoverable in the gui. We could also reduce the efficiency, if too many spawners are there, so their cummulative speed would stay the same. So if there are 11 spawner, max is 10, there should be a 10/11 chance to succeed the spawn. Also make that shown in the gui.
that's my opinion, but we could also discuss this

Rover656 replied:
I'd say an efficiency cap is reasonable to avoid overcrowding; should definitely be configurable and displayed in a tooltip for clarity.

ferriarnus replied:
Hmm, so simply make there be a chance for them to fail if to many are present? That should be fine, I can add it to the "blocked" enum.

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/task/SpawnTask.java:
```diff
@@ -0,0 +1,133 @@
+            double z = pos.getZ() + (randomsource.nextDouble() - randomsource.nextDouble()) * (double)this.blockEntity.getRange() + 0.5D;
+            if (level.noCollision(blockEntity.getEntityType().getAABB(x, y, z))) {
+
+                Entity entity = EntityType.loadEntityRecursive(blockEntity.getEntityData().getEntityTag(), level, entity1 -> {
```
should we allow exact copise? could result to dupes when entities hold shulkers or in general items

ferriarnus replied:
Hmm good point, maybe just getting the entitytype would be enough.

justliliandev replied:
maybe put it in a config, multiple copymodes like none, health and all and add a comment what each mode does

ferriarnus replied:
Make this an enum somewhere or how do you see it?

justliliandev replied:
enum sounds reasonable

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/task/SpawnTask.java:
```diff
@@ -0,0 +1,133 @@
+                level.levelEvent(2004, pos, 0);
+                level.gameEvent(entity, GameEvent.ENTITY_PLACE, new BlockPos(x, y, z));
+                if (entity instanceof Mob) {
+                    ((Mob)entity).spawnAnim();
```
use instanceof cast

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/task/SpawnTask.java:
```diff
@@ -0,0 +1,133 @@
+                }
+
+                if (entity instanceof Mob) {
+                    Mob mob = (Mob)entity;
```
use instanceof cast

justliliandev commented on src/machines/java/com/enderio/machines/common/item/PoweredSpawnerItem.java:
```diff
@@ -0,0 +1,42 @@
+    public void addCommonTooltips(ItemStack itemStack, @Nullable Player player, List<Component> tooltips) {
+        itemStack
+            .getCapability(EIOCapabilities.ENTITY_STORAGE)
+            .ifPresent(entityStorage -> entityStorage
```
would resolve and map calls work to prevent 2 nested ifPresents? if not, this is fine

ferriarnus replied:
I mean probably. Do you prefer it over this?

justliliandev replied:
map and flatmap would probably look nicer , but it's not important and personal preference

justliliandev commented on src/machines/java/com/enderio/machines/common/menu/PoweredSpawnerMenu.java:
```diff
@@ -0,0 +1,33 @@
+        super(blockEntity, inventory, MachineMenus.POWERED_SPAWNER.get(), pContainerId);
+
+        if (blockEntity != null) {
+            if (blockEntity.requiresCapacitor()) {
```
merge to one if statement

justliliandev commented on src/main/java/com/enderio/base/common/tag/EIOTags.java:
```diff
@@ -94,6 +94,7 @@ private static void init() {}
 
         public static final TagKey<Fluid> COLD_FIRE_IGNITER_FUEL = FluidTags.create(EnderIO.loc("fluid_fuel/cold_fire_igniter"));
         public static final TagKey<Fluid> STAFF_OF_LEVITY_FUEL = FluidTags.create(EnderIO.loc("fluid_fuel/staff_of_levity"));
+        public static final TagKey<Fluid> EXPERIENCE = FluidTags.create(new ResourceLocation("forge", "fluid/experience")); //TODO find tag?
```
the tag used by cyclic is `forge:experience`, we should use that as well

ferriarnus replied:
Hmm this is actually more complex maybe. If the tank has one type of experience fluid, will it allow another to be "merged" into it? Should we make this happen? The safest is to only use our own experience though.

justliliandev replied:
it would be the safest, but I think it should accept all of them

ferriarnus replied:
The thing is that it would need some special code and care, since it would need to ignore the (stack) fluid and still fill when it's not the same. And than we get to the issue of what to return inside the tank. *And possible mods using different units...*

justliliandev replied:
mods using different amounts for the same thing is not the case I think. And if so a standard would be better, because why should the tag exist if it's not meant for equality

ferriarnus replied:
I'll give it a go, but I might need to rewrite some stuff. And it will need some good testing.

justliliandev commented on src/main/java/com/enderio/base/common/util/ExperienceUtil.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public static Vector2i getLevelFromFluidWithLeftover(int fluidAmount, int startLevel, int stopLevel) {
+        return getLevelForExpWithLeftover(fluidAmount/24, startLevel, stopLevel);
```
put 24 in a private static final field

ferriarnus committed:
resolving comments + fixes

ferriarnus committed:
Update SpawnTask.java

fix energy consumption

ferriarnus committed:
add spawn type enum

dphaldes commented on src/machines/java/com/enderio/machines/common/blockentity/PoweredSpawnerBlockEntity.java:
```diff
@@ -0,0 +1,126 @@
+    private StoredEntityData entityData = StoredEntityData.empty();
+    private int range = 3;
+    private boolean rangeVisible;
+    protected float rCol = 1;
+    protected float gCol = 0;
+    protected float bCol = 0;
```
Should this be configurable through the config like Vacuum Blocks ?

ferriarnus replied:
I'm planning on bringing all the range methods to the base machine class in a next PR. Should these colours be in a config, they might have been in the past, but does it have an advantage? Does it even help colour blind people?

dphaldes replied:
The config right now resides on the server side and thus not very useful as the colors will be synced to all players. It can be helpful but we need to move it client side for any use.  If we do decide to not make it configurable, we will need to find a good default color that works for all

dphaldes commented on src/machines/java/com/enderio/machines/common/blockentity/SoulBinderBlockEntity.java:
```diff
@@ -0,0 +1,146 @@
+    public MachineInventoryLayout getInventoryLayout() {
+        return MachineInventoryLayout.builder()
+            .setStackLimit(1)
+            .inputSlot((slot, stack) -> stack.getCapability(EIOCapabilities.ENTITY_STORAGE).isPresent())
```
This capability is present in both the soul vial and broken spawner. The recipe only allows for soul vials in the first position

dphaldes replied:
You can also add Powered Spawners in the first slot which shouldn't happen

ferriarnus replied:
This is part of the design when I made it. It needs the first slot to have the capability for the recipe to work. It doesn't impose other restriction besides that, so yeah, powered spawners could also fit in it. I didn't want to limit the first slot to just soul vials, though maybe that's just simpler?

dphaldes replied:
The recipe checks for soul vial and so it probably shouldn't accept other items (since we don't a recipe validity notifier inside the screen). I can see people getting confused when items get accepted in slots but nothing happens.

ferriarnus replied:
I wanted to be future proof in case we did want a recipe that used something else. But now that I think about it, that's probably not gonna happen. I'll change it to only take soul vials, that's probably fine.

justliliandev commented on src/main/java/com/enderio/base/common/util/ExperienceUtil.java:
```diff
@@ -4,6 +4,17 @@
 
 public class ExperienceUtil {
 
+    // 1 exp = 24 mb
+    public static int EXPTOFLUID = 20;
```
either fix comment or fix number

rename to EXP_TO_FLUID

albinaask commented on src/machines/java/com/enderio/machines/common/recipe/SoulBindingRecipe.java:
```diff
@@ -0,0 +1,209 @@
+
+    public static class Container extends RecipeWrapper {
+
+        private final FluidTank fluidTank;
```
You may want to swap this for a machineFluidTank, since you only want fluid xp to flow in. You can do both of those things automatically through that class ;)

ferriarnus replied:
This container is purely used for the recipe and is just a reference to the real tank. But you do raise a good point of having a second look at the real tank, as it might be replaced by machineFluidTank. It also seems to miss somethings right now, good catch.

ferriarnus replied:
`machineFluidTank` isn't a thing though, as far as I can see. We do have a `MachineFluidHandler` is that what you ment?

ferriarnus committed:
push before the weekend

- fixed progress widget
- added components for blocked and entity

TODO
- Fluid tank

ferriarnus committed:
fix gui + filter tank

ferriarnus committed:
Update SoulBinderBlockEntity.java

ferriarnus committed:
align gui text

ferriarnus committed:
Merge branch '1.19.x' into feature/soulbinder

ferriarnus committed:
add proper translations

Rover656 commented on build.gradle:
```diff
@@ -190,6 +190,10 @@ repositories {
 
     maven { url 'https://maven.blamejared.com' }
 
+    // Temporary for Registrate
```
This can be fully removed now; if its necessary in future I'll put it back with the proper URL :)

Rover656 commented on src/machines/java/com/enderio/machines/common/blockentity/PoweredSpawnerBlockEntity.java:
```diff
@@ -0,0 +1,166 @@
+    public PoweredSpawnerBlockEntity(BlockEntityType type, BlockPos worldPosition, BlockState blockState) {
+        super(CAPACITY, TRANSFER, USAGE, type, worldPosition, blockState);
+        add2WayDataSlot(new BooleanDataSlot(this::isShowingRange, this::shouldShowRange, SyncMode.GUI));
+        addDataSlot(new ResourceLocationDataSlot(() -> this.getEntityType().orElse(new ResourceLocation("pig")),this::setEntityType, SyncMode.GUI));
```
Just a stylistic point; maybe store the RL of the pig in a constant just so its abundantly clear (and we're also not creating new objects regularly)

Rover656 commented on src/machines/java/com/enderio/machines/common/lang/MachineLang.java:
```diff
@@ -18,5 +18,11 @@ public class MachineLang {
 
     public static final Component TOOLTIP_ENERGY_EQUIVALENCE = REGISTRATE.addLang("gui", EnderIO.loc("energy_equivalence"), "A unit of energy, equivalent to FE.");
 
+    public static final Component TOO_MANY_MOB = REGISTRATE.addLang("gui", EnderIO.loc("spawner.too_many_mob"), "Too many Mobs");
```
Capitalisation of Mobs and Spawners is likely unnecessary here; but could be stylistic preference on my part.

Rover656 commented on src/main/java/com/enderio/base/common/config/common/BlocksConfig.java:
```diff
@@ -4,28 +4,19 @@
         builder.push("blocks");
 
-        builder.push("brokenSpawner");
+        builder.push("brokenSpawner"); //TODO is this push needed?
```
The clarity here is nice because otherwise it'd be a random `dropChance` key with no context.

Rover656 reviewed (changes requested):
Looking good, just leaving a few notes for you here :)

ferriarnus committed:
Merge branch '1.19.x' into feature/soulbinder

ferriarnus committed:
fix merge conflicts

Rover656 reviewed (approved):
I think this looks pretty good to go now, just need that model and we're golden! Nice work!

Rover656 committed:
Model for powered spawner and a PoweredTaskMachine fix

Rover656 committed:
Put them back

ferriarnus committed:
make it clearer no mob was found

justliliandev commented on src/machines/java/com/enderio/machines/client/gui/screen/PoweredSpawnerScreen.java:
```diff
@@ -27,25 +33,43 @@ public PoweredSpawnerScreen(PoweredSpawnerMenu pMenu, Inventory pPlayerInventory
+    public void render(PoseStack pPoseStack, int pMouseX, int pMouseY, float pPartialTicks) {
+        super.render(pPoseStack, pMouseX, pMouseY, pPartialTicks);
+        Optional<ResourceLocation> rl = getMenu().getBlockEntity().getEntityType();
+        if (rl.isPresent()) {
+            Optional<EntityType<?>> typeOptional = Registry.ENTITY_TYPE.getOptional(rl.get());
+            if (typeOptional.isPresent() && Registry.ENTITY_TYPE.getKey(typeOptional.get()).equals(rl.get())) { // check we don't get the default pig
+                String name = typeOptional.get().getDescription().getString();
+                font.draw(pPoseStack, name, leftPos + imageWidth/2f - font.width(name)/2f, topPos + 15, 0);
+            } else {
+                font.draw(pPoseStack, rl.get().toString(), leftPos + imageWidth/2f - font.width(rl.get().toString())/2f, topPos + 15, 0);
+            }
+        }
+        if (getMenu().getBlockEntity().getReason() != PoweredSpawnerBlockEntity.SpawnerBlockedReason.NONE) {
+            font.draw(pPoseStack, getMenu().getBlockEntity().getReason().getComponent().getString(), leftPos + imageWidth/2f - font.width(getMenu().getBlockEntity().getReason().getComponent().getString())/2f, topPos + 26, 0);
+        }
```
maybe do that in renderLabels

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/PoweredSpawnerBlockEntity.java:
```diff
@@ -117,10 +133,36 @@ public void shouldShowRange(Boolean show) {
-        if (!isClientSide() && level instanceof ServerLevel level) {
-            for (ServerPlayer player : level.players()) {
-                level.sendParticles(player, data, true, pos.x, pos.y, pos.z, 1, 0, 0, 0, 0);
+        if (!isClientSide() && level instanceof ServerLevel serverLevel) {
```
remove !isClientSide() check as that is already forced by ServerLevel

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/PoweredSpawnerBlockEntity.java:
```diff
@@ -117,10 +133,36 @@ public void shouldShowRange(Boolean show) {
+
+        private final Component component;
+
+        SpawnerBlockedReason(String name, Component component) {
```
remove unused name parameter

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/SoulBinderBlockEntity.java:
```diff
@@ -118,12 +119,20 @@ public int getNeededXP() {
                 setChanged();
             }
+
+            @Override
+            public int fill(FluidStack resource, FluidAction action) {
+                if (!this.getFluid().getFluid().equals(EIOFluids.XP_JUICE.getSource()) && this.isFluidValid(resource)) { // Auto convert to own fluid (source) type?
+                    resource = new FluidStack(EIOFluids.XP_JUICE.getSource(), resource.getAmount());
+                }
+                return super.fill(resource, action);
+            }
```
remove question mark and add the comment on the line above

ferriarnus replied:
It was since I wasn't sure we should do it. But if you're fine with it, that's good.

justliliandev replied:
yeah, I'm fine with that approach

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/SoulBinderBlockEntity.java:
```diff
@@ -118,12 +119,20 @@ public int getNeededXP() {
+
+            @Override
+            public int fill(FluidStack resource, FluidAction action) {
+                if (!this.getFluid().getFluid().equals(EIOFluids.XP_JUICE.getSource()) && this.isFluidValid(resource)) { // Auto convert to own fluid (source) type?
```
```suggestion
                if (!this.getFluid().getFluid().isSame(EIOFluids.XP_JUICE.getSource()) && this.isFluidValid(resource)) { // Auto convert to own fluid (source) type?
```

ferriarnus committed:
change sugestions

ferriarnus committed:
Why labels, why?

move render code to labels

ferriarnus committed:
Update PoweredSpawnerScreen.java

Rover656 reviewed (approved):
:shipit:
